notifier/notifier.go

```diff
@@ -221,43 +245,114 @@ func (n *Notifier) collectRogueCallbacks() {
+// kafka has already tried to retransmit a message for an amount of time. For http if the
+// callback is not successful we have already handled this case in Notify() method with retryOrFail().
+// Also not that when on http we call QueueJobForDeletion which schedules file deletion associated to our job,
+// because we are sure the other end has downloaded the file. But kakfa only ensures that a message has been
+// persisted, but most likely consumer has not recieved the message yet alone the file.
+func (n *Notifier) handleCallbackInfo(cbInfo job.Callback) error {
+	var e string
+
+	j, err := n.Storage.GetJob(cbInfo.JobID)
 	if err != nil {
-		return err
+		e = fmt.Sprintf("\nError: Could not get job %s. Operation returned error: %s", cbInfo.JobID, err)
+		return errors.New(e)
 	}
 
-	cbInfo, err := n.getCallbackInfo(j)
+	if cbInfo.Delivered {
+		cbType, _ := j.GetCallbackTypeAndDestination()
+
+		if cbType == backend.HTTP {
+			err := n.Storage.QueueJobForDeletion(cbInfo.JobID)
+			if err != nil {
+				e = fmt.Sprintf("\nError: Could not queue job %s for deletion: %s", cbInfo.JobID, err)
+				return errors.New(e)
+			}
+		}
+
+		n.stats.Add(statsSuccessfulCallbacks, 1)
+
+		err := n.Storage.RemoveJob(cbInfo.JobID)
+		if err != nil {
+			e = fmt.Sprintf("Could not remove job %s. Operation returned error: %s", cbInfo.JobID, err)
+			return errors.New(e)
+		}
+		return nil
+	}
+
+	err = n.markCbFailed(&j, cbInfo.Error)
 	if err != nil {
-		return n.markCbFailed(j, err.Error())
+
+		e = fmt.Sprintf("\nError during marking callback as failed for job %s. Operation returned error: %s", cbInfo.JobID, err)
+		return errors.New(e)
 	}
 
-	cb, err := json.Marshal(cbInfo)
+	return nil
+}
+
+// Prepare runs the necessary actions before calling Notify
+// and returns a callback info
+func (n *Notifier) Prepare(j *job.Job) (job.Callback, error) {
+	j.CallbackCount++
+
+	err := n.markCbInProgress(j)
 	if err != nil {
-		return n.markCbFailed(j, err.Error())
+		return job.Callback{}, err
 	}
 
-	n.Log.Println("Performing callback request for", j, "...")
-	res, err := n.client.Post(j.CallbackURL, "application/json", bytes.NewBuffer(cb))
-	if err != nil || res.StatusCode < 200 || res.StatusCode >= 300 {
-		if err == nil {
-			err = fmt.Errorf("Received Status: %s", res.Status)
-		}
-		return n.retryOrFail(j, err.Error())
+	cbInfo, err := j.CallbackInfo(*n.DownloadURL)
+	if err != nil {
+		return job.Callback{}, n.markCbFailed(j, err.Error())
 	}
 
-	if res.StatusCode == http.StatusAccepted {
-		err := n.Storage.QueueJobForDeletion(j.ID)
-		if err != nil {
-			n.Log.Println("Error: Could not queue job for deletion", err)
-		}
+	return cbInfo, nil
+}
+
+// Notify posts callback info to job's destination by calling Notify
+// on each backend.
+func (n *Notifier) Notify(j *job.Job, cbInfo job.Callback) error {
+	n.Log.Println("Performing callback action for", j, "...")
+
+	cbType, cbDst := j.GetCallbackTypeAndDestination()
```

Comment:
Since we have `j` and `cbInfo`, why do we still need to call `GetCallbackTypeAndDestination()`?

Reply:
Until callback_url is completely removed we need to be consistent with old and new approach. We discussed this in person.